juju/juju
# Store machineID on subordinate unit document.

Subordinate units are created when the principal unit enters the relation scope. At this time, the machine ID is known and never changes. Rather than have the subordinate unit always going throught the principal unit document to work out the assigned machine ID, have the subordinate document created with the machine ID from the principal.

There was quite a bit of test fallout with this change as many of the tests did not follow the real world and didn't bother assigning the principal unit to a machine before creating the subordinate. These tests have been updated.

One of the primary reasons for doing this piece of work is to allow the allwatcher to be able to process the initial units of a model without additional queries per unit. Before this change a principal unit would have read the machine document twice, once for the public address, and once for the private address. A subordinate would have read the principal document twice in addition to the machine document. The allwatcher change now stores the public and private address on the core/multiwatcher type, and the unit handling now doesn't need additional queries to get the machine ID. CAAS units still get the addresses directly from the unit itself as the logic there is more complicated and deals with entities that aren't yet stored in the multiwatcher store.

## QA steps

```sh
juju deploy ~jameinel/ubuntu-lite
juju deploy telegraf
juju relate ubuntu-lite telegraf
```
Observe that the telegraf unit gets an address in juju status.

## Documentation changes

Internal change only.
